rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/model/ParametricFileFacts.java

```diff
@@ -110,64 +138,175 @@ public void close(ISourceLocation loc) {
-        private volatile List<Diagnostic> typeCheckerMessages = Collections.emptyList();
-        private final ParametricSummaryBridge summary;
+
+        // To replace old diagnostics when new diagnostics become available in a
+        // thread-safe way, we need to atomically: (1) check if the version of
+        // the new diagnostics is greater than the version of the old
+        // diagnostics; (2) if so, replace old with new. This is why
+        // `AtomicReference` and `Versioned` are needed in the following three
+        // fields.
+        private final AtomicReference<Versioned<List<Diagnostic>>> parserDiagnostics = Versioned.atomic(-1, Collections.emptyList());
+        private final AtomicReference<Versioned<List<Diagnostic>>> analyzerDiagnostics = Versioned.atomic(-1, Collections.emptyList());
+        private final AtomicReference<Versioned<List<Diagnostic>>> builderDiagnostics = Versioned.atomic(-1, Collections.emptyList());
+
+        private final ParametricSummaryBridge analyzer;
+        private final ParametricSummaryBridge builder;
+
+        private final AtomicInteger latestVersionCalculateAnalyzer = new AtomicInteger();
+
+        @SuppressWarnings("java:S3077") // Reads/writes happen sequentially
+        private volatile @MonotonicNonNull InterruptibleFuture<Lazy<List<Diagnostic>>> builderAnalysis = null;
+        @SuppressWarnings("java:S3077") // Reads/writes happen sequentially
+        private volatile @MonotonicNonNull InterruptibleFuture<Lazy<List<Diagnostic>>> builderBuild = null;
 
         public FileFact(ISourceLocation file) {
             this.file = file;
-            this.summary = new ParametricSummaryBridge(exec, file, columns, contrib, lookupState);
+            this.analyzer = new ParametricSummaryBridge(exec, file, columns, contrib, lookupState, contrib::analyze);
+            this.builder = new ParametricSummaryBridge(exec, file, columns, contrib, lookupState, contrib::build);
         }
 
         public void reloadContributions() {
-            summary.reloadContributions();
+            analyzer.reloadContributions();
+            builder.reloadContributions();
+        }
+
+        private <T> void reportDiagnostics(AtomicReference<Versioned<T>> current, int version, T messages) {
+            var maybeNewer = new Versioned<>(version, messages);
+            if (Versioned.replaceIfNewer(current, maybeNewer)) {
+                sendDiagnostics();
+            }
+        }
+
+        public void invalidateAnalyzer(boolean isClosing) {
+            analyzer.invalidate(isClosing);
+        }
+
+        public void invalidateBuilder(boolean isClosing) {
+            builder.invalidate(isClosing);
+            if (builderAnalysis != null) {
+                builderAnalysis.interrupt();
+            }
+            if (builderBuild != null) {
+                builderBuild.interrupt(); // Sets the interrupt flag in `builderBuild`, which `builder.invalidate` doesn't do
```

Comment:
Why do we need this? why is `builder.invalidate` not enough (or can we make it be enough?).